klt-cert-manager/controllers/keptnwebhookcontroller/resource_retriever.go

```diff
@@ -0,0 +1,118 @@
+	var mutatingWebhooks admissionregistrationv1.MutatingWebhookConfigurationList
+
+	if err := r.Client.List(ctx, &mutatingWebhooks, client.MatchingLabels(r.MatchLabels)); err != nil {
+		return nil, err
+	}
+	return &mutatingWebhooks, nil
+}
+
+func (r LabelSelectorRetriever) GetValidatingWebhooks(ctx context.Context) (*admissionregistrationv1.ValidatingWebhookConfigurationList, error) {
+	var validatingWebhooks admissionregistrationv1.ValidatingWebhookConfigurationList
+
+	if err := r.Client.List(ctx, &validatingWebhooks, client.MatchingLabels(r.MatchLabels)); err != nil {
+		return nil, err
+	}
+	return &validatingWebhooks, nil
+}
+
+func (r LabelSelectorRetriever) GetCRDs(ctx context.Context) (*apiv1.CustomResourceDefinitionList, error) {
+	var crds apiv1.CustomResourceDefinitionList
+	opt := client.MatchingLabels(r.MatchLabels)
+	if err := r.Client.List(ctx, &crds, opt); err != nil {
+		return nil, err
+	}
+
+	return &crds, nil
+}
+
+type ResourceNameRetriever struct {
+	Client         client.Client
+	WatchResources ObservedObjects
+	Log            logr.Logger
+}
+
+func (r ResourceNameRetriever) GetMutatingWebhooks(ctx context.Context) (*admissionregistrationv1.MutatingWebhookConfigurationList, error) {
+	result := &admissionregistrationv1.MutatingWebhookConfigurationList{
+		Items: []admissionregistrationv1.MutatingWebhookConfiguration{},
+	}
+	for _, mwhName := range r.WatchResources.MutatingWebhooks {
+		mwh := &admissionregistrationv1.MutatingWebhookConfiguration{}
+		if err := r.Client.Get(ctx, client.ObjectKey{Name: mwhName}, mwh); err != nil {
+			r.Log.Error(err, "Could not retrieve MutatingWebhookConfiguration", "MutatingWebhookConfiguration", mwhName)
+		} else {
+			result.Items = append(result.Items, *mwh)
+		}
+	}
+
+	return result, nil
+}
+
+func (r ResourceNameRetriever) GetValidatingWebhooks(ctx context.Context) (*admissionregistrationv1.ValidatingWebhookConfigurationList, error) {
+	result := &admissionregistrationv1.ValidatingWebhookConfigurationList{
+		Items: []admissionregistrationv1.ValidatingWebhookConfiguration{},
+	}
+	for _, vwhName := range r.WatchResources.ValidatingWebhooks {
+		vwh := &admissionregistrationv1.ValidatingWebhookConfiguration{}
+		if err := r.Client.Get(ctx, client.ObjectKey{Name: vwhName}, vwh); err != nil {
+			r.Log.Error(err, "Could not retrieve ValidatingWebhookConfiguration", "ValidatingWebhookConfiguration", vwhName)
+		} else {
+			result.Items = append(result.Items, *vwh)
+		}
+	}
+
+	return result, nil
+}
+
+func (r ResourceNameRetriever) GetCRDs(ctx context.Context) (*apiv1.CustomResourceDefinitionList, error) {
+	result := &apiv1.CustomResourceDefinitionList{
+		Items: []apiv1.CustomResourceDefinition{},
+	}
+	for _, crdName := range r.WatchResources.CustomResourceDefinitions {
+		crd := &apiv1.CustomResourceDefinition{}
+		if err := r.Client.Get(ctx, client.ObjectKey{Name: crdName}, crd); err != nil {
```

Comment:
We could, although then the operator using this library would then need read permissions for all webhooks/CRDs in the cluster - which I believe should be ok, since it would still be possible to restrict the update permissions to exactly the resources you want to be managed by the operator.
Another reason why I went with GET instead of LIST here though, is that it's more likely to have a large total number of CRDs in the whole cluster, but only a relatively small number that should be managed by a certain operator. Also the reconciliation loop of the cert controller is triggered infrequently (only at changes of the observed deployments, or every 3 hours to check if the certificate is still valid), so therefore the performance impact will not be high.

Reply:
I got the point thanks for explanation, I am good with going with the current solution